File: 3217.delete-nodes-from-linked-list-present-in-array.py
#
# @lc app=leetcode id=3217 lang=python3
#
# [3217] Delete Nodes From Linked List Present in Array
#
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        nums_set = set(nums)  # Convert nums array to set for O(1) lookups
        dummy = ListNode(0)  # Create a dummy node to handle edge cases easily
        dummy.next = head
        prev = dummy  # Initialize prev as dummy node
        current = head  # Start with the head of the linked list
        while current:
            if current.val in nums_set:  # If current node's value is in nums_set
                prev.next = current.next  # Skip the current node by linking prev to current's next
            else:
                prev = current  # Move prev pointer forward only if we didn't remove current node
            current = current.next  # Move to next node in the list
        return dummy.next  # Return the modified list starting from dummy's next node
# @lc code=end

File: 3217.delete-nodes-from-linked-list-present-in-array.java
#
# @lc app=leetcode id=3217 lang=java
#
# [3217] Delete Nodes From Linked List Present in Array
#
# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
import java.util.HashSet;
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        HashSet<Integer> numSet = new HashSet<>();
        for (int num : nums) { 
            numSet.add(num); 
        } 
        
        ListNode dummy = new ListNode(0); 
        dummy.next = head; 
        ListNode prev = dummy; 
        
        while (head != null) { 
            if (numSet.contains(head.val)) { 
                prev.next = head.next; 
            } else { 
                prev = head; 
            } 
            head = head.next; 
        } 
        return dummy.next; 
    } 
}
n# @lc code=end